File: tianshu/core/agent/runtime.py
from __future__ import annotations

import asyncio
import json
import re
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Coroutine

from tianshu.core.llm.base import LLMMessage, ToolCall
from tianshu.core.llm.factory import create_provider
from tianshu.core.log import get_logger
from tianshu.core.review.system import ReviewSystem, gate_tool
from tianshu.core.tools.registry import ToolRegistry, Tool

logger = get_logger("agent.runtime")


@dataclass
class ToolResult:
    name: str
    output: str
    error: str | None = None


@dataclass
class AgentResult:
    content: str
    tool_calls: list[ToolResult] = field(default_factory=list)
    child_agents: list[str] = field(default_factory=list)
    error: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "content": self.content,
            "tool_calls": [t.__dict__ for t in self.tool_calls],
            "child_agents": self.child_agents,
            "error": self.error,
        }


class MessageBus:
    """Agent 之间的消息总线,支持任意 Agent 相互调用与等待回复。"""

    def __init__(self) -> None:
        self._agents: dict[str, "Agent"] = {}
        self._pending: dict[str, asyncio.Future[AgentResult]] = {}
        self._lock = asyncio.Lock()

    def register(self, agent: "Agent") -> None:
        self._agents[agent.name] = agent

    def unregister(self, name: str) -> None:
        self._agents.pop(name, None)

    async def send(self, to: str, message: str, sender: str = "") -> AgentResult:
        agent = self._agents.get(to)
        if agent is None:
            raise KeyError(f"目标 Agent 不存在: {to}")
        return await agent.handle_message(message, sender)


class Agent:
    def __init__(
        self,
        name: str,
        system_prompt: str,
        provider_name: str,
        model: str = "",
        base_url: str = "",
        api_key: str = "",
        registry: ToolRegistry | None = None,
        bus: MessageBus | None = None,
        review: ReviewSystem | None = None,
        temperature: float = 0.2,
        max_tokens: int | None = None,
        max_iterations: int = 10,
        debug: bool = False,
        event_sink: Callable[[str, str, dict], Awaitable[None]] | None = None,
        cancelled: asyncio.Event | None = None,
    ) -> None:
        self.name = name
        self.system_prompt = system_prompt
        self.registry = registry or ToolRegistry()
        self.bus = bus or MessageBus()
        self.review = review
        self.provider = create_provider(provider_name, base_url, model, api_key, temperature=temperature, max_tokens=max_tokens)
        self.max_iterations = max_iterations
        self.messages: list[LLMMessage] = []
        self.debug = debug
        self.event_sink = event_sink
        self.cancelled = cancelled or asyncio.Event()

    async def handle_message(self, message: str, sender: str = "") -> AgentResult:
        history = [LLMMessage(role="system", content=self.system_prompt)]
        history += self.messages
        history.append(LLMMessage(role="user", content=f"(来自 {sender or '上级'} 的任务)\n{message}"))
        try:
            return await self._run_loop(history)
        except Exception as e:  # noqa: BLE001
            logger.exception("Agent %s 运行异常", self.name)
            return AgentResult(content="", error=str(e))

    async def _run_loop(self, history: list[LLMMessage]) -> AgentResult:
        tool_results: list[ToolResult] = []
        children: list[str] = []
        tools = self.registry.schemas()

        for i in range(self.max_iterations):
            result = await self.provider.chat(history, tools=tools, cancel_event=self.cancelled)
            history.append(LLMMessage(role="assistant", content=result.content or "", tool_calls=result.tool_calls))

            if not result.has_tool_calls:
                self.messages = history[1:]
                return AgentResult(content=result.content or "", tool_calls=tool_results, child_agents=children)

            if self.cancelled.is_set():
                return AgentResult(content="(任务已取消)", tool_calls=tool_results, child_agents=children)

            for tc in result.tool_calls:
                if self.cancelled.is_set():
                    return AgentResult(content="(任务已取消)", tool_calls=tool_results, child_agents=children)
                if tc.name == "call_agent":
                    child_name = tc.arguments.get("agent", "")
                    await self._emit("agent_action", agent=self.name, action="call_agent", target=child_name)
                    child = await self._dispatch_child(tc)
                    children.append(child_name)
                    output = child.content or "(无输出)"
                    if child.error:
                        output = f"错误: {child.error}"
                else:
                    await self._emit("agent_action", agent=self.name, action="tool", tool=tc.name, args=tc.arguments)
                    output, err = await self._exec_tool(tc)
                    tool_results.append(ToolResult(name=tc.name, output=output or "", error=err))
                    if err:
                        output = f"错误: {err}"
                history.append(
                    LLMMessage(role="tool", tool_call_id=tc.id, content=str(output)[:12000])
                )

        self.messages = history[1:]
        return AgentResult(content="(达到最大迭代次数)", tool_calls=tool_results, child_agents=children)

    async def _emit(self, event: str, **data: Any) -> None:
        if self.event_sink:
            try:
                await self.event_sink(self.name, event, data)
            except Exception:  # noqa: BLE001
                logger.exception("event_sink 异常 agent=%s event=%s", self.name, event)

    async def _dispatch_child(self, tc: ToolCall) -> AgentResult:
        name = tc.arguments.get("agent", "")
        task = tc.arguments.get("task", "")
        return await self.bus.send(name, task, sender=self.name)

    async def _exec_tool(self, tc: ToolCall) -> tuple[str, str | None]:
        tool = self.registry.get(tc.name)
        if tool is None:
            return "", f"工具不存在: {tc.name}"
        try:
            if tool.requires_review:
                if self.review is None:
                    return "", f"工具 {tc.name} 需要审核但未配置审核系统"
                await gate_tool(self.review, self.name, tool, tc.arguments)
            out = tool.func(**tc.arguments)
            if asyncio.iscoroutine(out):
                out = await out
            if isinstance(out, str):
                return out, None
            if tool.format_result == "json":
                return json.dumps(out, ensure_ascii=False, default=str), None
            return str(out), None
        except Exception as e:  # noqa: BLE001
            return "", f"{type(e).__name__}: {e}"


def build_agent_call_tool(bus: MessageBus) -> Tool:
    async def call_agent(agent: str, task: str) -> str:
        result = await bus.send(agent, task)
        return result.content or "(无输出)"

    return Tool(
        name="call_agent",
        description="调用另一个 Agent 处理子任务,传入目标 Agent 名称和任务描述",
        func=call_agent,
        parameters={
            "type": "object",
            "properties": {
                "agent": {"type": "string", "description": "目标 Agent 名称"},
                "task": {"type": "string", "description": "分配给目标 Agent 的任务"},
            },
            "required": ["agent", "task"],
        },
    )

File: tiankui-ts/src/core/agent/runtime.ts
import { CancellationToken } from "../llm/types.js";
import type { LLMMessage, ToolCall } from "../llm/types.js";
import { createProvider } from "../llm/factory.js";
import { gateTool } from "../review/system.js";
import type { ReviewSystem } from "../review/system.js";
import { Tool, ToolRegistry } from "../tools/registry.js";

export interface ToolResult {
  name: string;
  output: string;
  error: string | null;
}

export interface AgentResult {
  content: string;
  tool_calls: ToolResult[];
  child_agents: string[];
  error: string | null;
}

export type EventSink = (agent: string, event: string, data: Record<string, unknown>) => Promise<void>;

export class MessageBus {
  private _agents = new Map<string, Agent>();

  register(agent: Agent): void {
    this._agents.set(agent.name, agent);
  }

  unregister(name: string): void {
    this._agents.delete(name);
  }

  names(): string[] {
    return [...this._agents.keys()];
  }

  agent(name: string): Agent | null {
    return this._agents.get(name) ?? null;
  }

  async send(to: string, message: string, sender = ""): Promise<AgentResult> {
    const agent = this._agents.get(to);
    if (!agent) throw new Error(`目标 Agent 不存在: ${to}`);
    return agent.handleMessage(message, sender);
  }
}

export interface AgentOptions {
  name: string;
  system_prompt: string;
  provider_name: string;
  model?: string;
  base_url?: string;
  api_key?: string;
  registry?: ToolRegistry;
  bus?: MessageBus;
  review?: ReviewSystem | null;
  temperature?: number;
  max_tokens?: number | null;
  max_iterations?: number;
  debug?: boolean;
  event_sink?: EventSink | null;
  cancelled?: CancellationToken | null;
}

export class Agent {
  name: string;
  system_prompt: string;
  registry: ToolRegistry;
  bus: MessageBus;
  review: ReviewSystem | null;
  provider: ReturnType<typeof createProvider>;
  max_iterations: number;
  messages: LLMMessage[] = [];
  debug: boolean;
  event_sink: EventSink | null;
  cancelled: CancellationToken;

  constructor(opts: AgentOptions) {
    this.name = opts.name;
    this.system_prompt = opts.system_prompt;
    this.registry = opts.registry ?? new ToolRegistry();
    this.bus = opts.bus ?? new MessageBus();
    this.review = opts.review ?? null;
    this.provider = createProvider(opts.provider_name, opts.base_url ?? "", opts.model ?? "", opts.api_key ?? "", {
      temperature: opts.temperature ?? 0.2,
      max_tokens: opts.max_tokens ?? null,
    });
    this.max_iterations = opts.max_iterations ?? 10;
    this.debug = opts.debug ?? false;
    this.event_sink = opts.event_sink ?? null;
    this.cancelled = opts.cancelled ?? new CancellationToken();
  }

  async handleMessage(message: string, sender = ""): Promise<AgentResult> {
    const history: LLMMessage[] = [{ role: "system", content: this.system_prompt }];
    history.push(...this.messages);
    history.push({ role: "user", content: `(来自 ${sender || "上级"} 的任务)\n${message}` });
    try {
      return await this._runLoop(history);
    } catch (e) {
      console.error(`Agent ${this.name} 运行异常`, e);
      return { content: "", tool_calls: [], child_agents: [], error: String(e) };
    }
  }

  private async _runLoop(history: LLMMessage[]): Promise<AgentResult> {
    const toolResults: ToolResult[] = [];
    const children: string[] = [];
    const tools = this.registry.schemas();
    const seenCalls = new Set<string>();

    for (let i = 0; i < this.max_iterations; i++) {
      const result = await this.provider.chat(history, { tools, cancel_event: this.cancelled });
      if (result.reasoning) {
        await this._emit("thinking", { agent: this.name, content: result.reasoning });
      }
      history.push({ role: "assistant", content: result.content ?? "", tool_calls: result.tool_calls });

      if (!result.tool_calls?.length) {
        this.messages = history.slice(1);
        return { content: result.content ?? "", tool_calls: toolResults, child_agents: children, error: null };
      }

      if (this.cancelled.isSet()) {
        return { content: "(任务已取消)", tool_calls: toolResults, child_agents: children, error: null };
      }

      for (const tc of result.tool_calls) {
        if (this.cancelled.isSet()) {
          return { content: "(任务已取消)", tool_calls: toolResults, child_agents: children, error: null };
        }
        const callKey = `${tc.name}:${JSON.stringify(Object.entries(tc.arguments).sort())}`;
        if (seenCalls.has(callKey)) {
          const output = `错误: 工具 ${tc.name} 与之前的调用参数完全相同,疑似死循环,已拦截(重复调用:${tc.name})`;
          toolResults.push({ name: tc.name, output: "", error: output });
          history.push({ role: "tool", tool_call_id: tc.id, content: output });
          continue;
        }
        seenCalls.add(callKey);
        let output: string;
        if (tc.name === "call_agent") {
          const childName = String(tc.arguments.agent ?? "");
          await this._emit("agent_action", { agent: this.name, action: "call_agent", target: childName });
          const child = await this._dispatchChild(tc);
          children.push(childName);
          output = child.content || "(无输出)";
          if (child.error) output = `错误: ${child.error}`;
        } else {
          await this._emit("agent_action", { agent: this.name, action: "tool", tool: tc.name, args: tc.arguments });
          const { output: out, err } = await this._execTool(tc);
          toolResults.push({ name: tc.name, output: out ?? "", error: err });
          output = err ? `错误: ${err}` : (out ?? "");
        }
        history.push({ role: "tool", tool_call_id: tc.id, content: String(output).slice(0, 12000) });
      }
    }

    this.messages = history.slice(1);
    return { content: "(达到最大迭代次数)", tool_calls: toolResults, child_agents: children, error: null };
  }

  private async _emit(event: string, data: Record<string, unknown>): Promise<void> {
    if (!this.event_sink) return;
    try {
      await this.event_sink(this.name, event, data);
    } catch {
      console.error(`event_sink 异常 agent=${this.name} event=${event}`);
    }
  }

  private async _dispatchChild(tc: ToolCall): Promise<AgentResult> {
    const name = String(tc.arguments.agent ?? "");
    const task = String(tc.arguments.task ?? "");
    return this.bus.send(name, task, this.name);
  }

  private async _execTool(tc: ToolCall): Promise<{ output: string; err: string | null }> {
    const tool = this.registry.get(tc.name);
    if (!tool) return { output: "", err: `工具不存在: ${tc.name}` };
    try {
      if (tool.requires_review) {
        if (!this.review) return { output: "", err: `工具 ${tc.name} 需要审核但未配置审核系统` };
        await gateTool(this.review, this.name, tool, tc.arguments);
      }
      const out = await tool.func(tc.arguments);
      if (typeof out === "string") return { output: out, err: null };
      if (tool.format_result === "json") {
        return { output: JSON.stringify(out), err: null };
      }
      return { output: String(out), err: null };
    } catch (e) {
      return { output: "", err: `${(e as Error).constructor.name}: ${(e as Error).message}` };
    }
  }
}

export function buildAgentCallTool(bus: MessageBus): Tool {
  const callAgent = async (args: Record<string, unknown>): Promise<string> => {
    const result = await bus.send(String(args.agent), String(args.task));
    return result.content || "(无输出)";
  };
  return new Tool({
    name: "call_agent",
    description: "调用另一个 Agent 处理子任务,传入目标 Agent 名称和任务描述",
    func: callAgent,
    parameters: {
      type: "object",
      properties: {
        agent: { type: "string", description: "目标 Agent 名称" },
        task: { type: "string", description: "分配给目标 Agent 的任务" },
      },
      required: ["agent", "task"],
    },
  });
}
